src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,865 @@
+        }
+      }
+      currentLayer->ComputeOutputDimensions();
+    }
+    else
+    {
+      exploring.push(std::make_pair(currentLayer, true));
+      for (size_t i = 0; i < numParents; i++)
+      {
+        exploring.push(std::make_pair(adjacencyList[currentLayer][i], false));
+      }
+    }
+  }
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::UpdateDimensions(const std::string& functionName,
+                    const size_t inputDimensionality)
+{
+  if (inputDimensions.size() == 0)
+    inputDimensions = { inputDimensionality };
+
+  size_t totalInputSize = 1;
+  for (size_t i = 0; i < inputDimensions.size(); i++)
+    totalInputSize *= inputDimensions[i];
+
+  if (totalInputSize != inputDimensionality && inputDimensionality != 0)
+  {
+    throw std::logic_error(functionName + ": input size does not match expected size set with InputDimensions()!");
+  }
+
+  ComputeOutputDimensions();
+  validOutputDimensions = true;
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+size_t DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::WeightSize()
+{
+  UpdateDimensions("DAGNetwork::WeightSize()");
+
+  size_t total = 0;
+  for (size_t i = 0; i < sortedNetwork.size(); i++)
+    total += sortedNetwork[i]->WeightSize();
+  return total;
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Reset(const size_t inputDimensionality)
+{
+  parameters.clear();
+
+  if (inputDimensionality != 0)
+  {
+    CheckNetwork("DAGNetwork::Reset()", inputDimensionality, true, false);
+  }
+  else if (inputDimensions.size() > 0)
+  {
+    size_t inputDim = inputDimensions[0];
+    for (size_t i = 1; i < inputDimensions.size(); i++)
+      inputDim *= inputDimensions[i];
+    CheckNetwork("DAGNetwork::Reset()", inputDim, true, false);
+  }
+  else
+  {
+    throw std::invalid_argument("DAGNetwork::Reset(): cannot reset network when no "
+        "input dimensionality is given, and `InputDimensions()` has not been "
+        "set!");
+  }
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::SetWeights(const MatType& weightsIn)
+{
+  if (!graphIsSet)
+    CheckGraph();
+
+  size_t offset = 0;
+  const size_t totalWeightSize = WeightSize();
+  for (size_t i = 0; i < sortedNetwork.size(); i++)
+  {
+    const size_t weightSize = sortedNetwork[i]->WeightSize();
+
+    if (offset + weightSize > totalWeightSize)
+    {
+      throw std::logic_error("DAGNetwork::SetLayerMemory(): Parameter size does not match total layer "
+        "weight size!");
```

Comment:
Here, too, is another instance where I think it would be a good idea to actually print what `weightSize` and `totalWeightSize` are.